feat: Redirect Blacksmith Sell/Scrap Messages to NotificationManager Overlay

san7890 opened the pull request:
Hey there,

I missed this from Wynntils Legacy, so I just decided to spend time to get it ported over from there. Has essentially the same features as the old version, although with a few new things:

Now relies completely on Regex Groups, as opposed to splitting it out based on the Section Character (don't have to toy around with sentence fragments).
Now handles the message when you sell/scrap multiple items to be singular/plural instead of just relying on an `item(s)` crutch (in the photographs below).

![image](https://user-images.githubusercontent.com/34697715/204160184-cd41d0e2-78e5-42e5-84aa-ecb6a0d22b55.png)

![image](https://user-images.githubusercontent.com/34697715/204160187-f949443b-b9b8-453a-88ae-c88013dc09e4.png)

![image](https://user-images.githubusercontent.com/34697715/204160190-ce9af117-a4da-4dc4-b186-e56ca16b8d4e.png)

![image](https://user-images.githubusercontent.com/34697715/204160196-189a3b68-c1f2-4a69-b935-ea0792af57bc.png)

san7890 committed:
framework

san7890 committed:
spotless

san7890 committed:
regex update, add to feature registries

san7890 committed:
bleaugh lets try it this way

san7890 committed:
yellow means go

san7890 committed:
another variation

san7890 committed:
actually implements the counter

this java thing hurts my brain

san7890 committed:
okay let's try an early return because of crashes

bleaugh

san7890 committed:
spotless time

san7890 committed:
wait maybe stripping formatting was the issue

san7890 committed:
charat is 0 apparently

san7890 committed:
last iterance of self-work

san7890 committed:
bolyai fixes

san7890 committed:
fingers crossed

san7890 committed:
gracefully handle single item selling

would drive me mad if i didn't do this one

san7890 committed:
better pluralizer

san7890 committed:
final touches, remove debug

san7890 committed:
Spotless Apply

san7890 committed:
typo

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,102 @@
+        // If we get emeralds, this will be the number of emeralds we get. If we get scrap, this will be the number of
+        // scrap we get.
+        String paymentString = "";
+        Matcher messageMatcher = BLACKSMITH_MESSAGE_PATTERN.matcher(event.getCodedMessage());
```
This check should be the first thing in the event

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        // The final part of the message.
+        {
```
Why the extra block? Remove it

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,102 @@
+            String sendableMessageString = sendableMessage.toString();
+
+            // Finally, we send the message.
+            NotificationManager.queueMessage(sendableMessageString);
```
I am not happy about this whole string generation part, it is a bit of a mess. 

Let's refactor this:
1, Create a StringBuilder to create the sold items (`1/0/0/0/0` part)
2, If we sold items -> "Sold %d (%s) item(s) for %dE",
    otherwise -> "Scrapped %d (%s) item(s) for %dscrap",

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -6,6 +6,7 @@
   "core.wynntils.category.redirects": "Game Notification Redirects",
   "core.wynntils.category.tooltips": "Item Tooltips",
   "core.wynntils.category.uncategorized": "Uncategorized",
+  "feature.wynntils.BlacksmithRedirectFeature.name": "Blacksmith Message Redirect",
```
This should not even run in dev env.
```suggestion
  "feature.wynntils.blacksmithRedirect.name": "Blacksmith Message Redirect",
```

kristofbolyai commented:
Also, why don't we include item count for scrapped items?  Either include it or only create the item count stringbuilder if we sold the items so we don't waste resources

san7890 committed:
Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

san7890 committed:
ci: spotless formatting

san7890 committed:
simple code review

san7890 committed:
Refactor to cut down on unused code

Keeps selling and scrapping in two separate "lanes" to reduce wasted resources.

Cuts down on stringbuilder excess.

san7890 committed:
spotless

san7890 committed:
Merge branch 'blacksmithery' of https://github.com/san7890/Artemis-bruh into blacksmithery

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,112 @@
+                countByTier.append(ChatFormatting.LIGHT_PURPLE);
+            }
+            countByTier.append(")");
+            countByTier.setCharAt(0, '(');
```
Add these in the format string below

san7890 replied:
This is what cleans it up from:

`/0/0/0/0/0/0/0/0/`

to:

`(0/0/0/0/0/0/0/0)`

Not sure how I would go about adding that to the format sting.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,112 @@
+            // Sold 1 (1/0/0/0/0/0/0/0) item for 4e.
+            sendableMessage = String.format(
+                    "%sSold %d %s %s for %s%s%s%s.",
+                    ChatFormatting.LIGHT_PURPLE,
```
You can include colors int the format string itself to make this look less weird

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,112 @@
+                    ChatFormatting.GREEN,
+                    paymentString,
+                    EmeraldSymbols.EMERALDS,
+                    ChatFormatting.LIGHT_PURPLE);
```
same

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,112 @@
+            String itemPluralizer = totalItemInteger == 1 ? "" : "s";
+
+            sendableMessage = String.format(
+                    "%sScrapped %d %s for %s%s scrap%s.",
```
Actually, is there a reason why we  don't have item tier breakdown here? Even if legacy didnt do it, I think it is nice to have it

san7890 replied:
Feels more cluttered to me IMO, but I can look into adding it. The only technical concern I can think of it the message overwriting it's bounds, like something I ran into where you might see:

"Scrapped 5 items for 16 scrap."
versus
"Scrapped 5 (0/5/0/0/0/0/0/0) items for
16 scrap"

might not matter though and we could be fine, the only concern is the relative fingerprint the sendablemessage will have

joshieman06 replied:
could ignore 0s (so instead of 0/5/0/0/0/0/0/1 it would be 5/1 but color coded)

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java:
```diff
@@ -0,0 +1,112 @@
+            // Build up the string that outlines how many items were sold in what tier (0/0/0/0/0/0/0/0).
+            StringBuilder countByTier = new StringBuilder();
+            for (ItemTier tier : ItemTier.values()) {
+                countByTier.append('/' + tier.getChatFormatting().toString() + totalItems.getOrDefault(tier, 0));
```
Use chained append here, Using + defeats the purpose of using stringbuilder

kristofbolyai reviewed (commented):
much better now! A few changes and it will be ready

san7890 committed:
review

removes chatformatting macros
patches up items glitch
patches up integer glitch

san7890 committed:
even moar spotless

san7890 committed:
uuuuuuuuuuuuuuuugh

san7890 committed:
palantir over ideaj

san7890 committed:
ci: spotless formatting

san7890 committed:
remove debugging code

san7890 committed:
code cleanup

kristofbolyai committed:
fix: Change matching to work with item rarities

kristofbolyai committed:
fix: This now fully works

kristofbolyai committed:
Merge branch 'alpha' into blacksmithery

san7890 commented:
Thank you!

kristofbolyai commented:
@Incompleteusern @magicus can one of you review this, as I have contributed to it too?
